module/zfs/arc.c

```diff
@@ -8553,15 +8859,21 @@ l2arc_add_vdev(spa_t *spa, vdev_t *vd)
 	/*
 	 * Create a new l2arc device entry.
 	 */
-	adddev = kmem_zalloc(sizeof (l2arc_dev_t), KM_SLEEP);
+	adddev = vmem_zalloc(sizeof (l2arc_dev_t), KM_SLEEP);
```

Comment:
I tried but with kmem_zalloc I get kernel panics:
```
[58361.800914] Large kmem_alloc(131272, 0x1000), please file an issue at:
                https://github.com/zfsonlinux/zfs/issues/new
[58361.800916] CPU: 2 PID: 252829 Comm: lt-zpool Tainted: P           OE     5.4.0-1-vfio-debug-00015-g777     6c0538c53 #1
[58361.800917] Hardware name: QEMU Standard PC (Q35 + ICH9, 2009), BIOS 0.0.0 02/06/2015
[58361.800917] Call Trace:
[58361.800921]  dump_stack+0x66/0x90
[58361.800925]  spl_kmem_zalloc.cold+0x19/0x1e [spl]
[58361.800948]  l2arc_add_vdev+0x2c/0x1e0 [zfs]
[58361.800976]  spa_load_l2cache+0x3db/0x500 [zfs]
[58361.801004]  spa_import+0x45c/0x870 [zfs]
[58361.801034]  zfs_ioc_pool_import+0x140/0x160 [zfs]
[58361.801065]  zfsdev_ioctl_common+0x71b/0x8f0 [zfs]
[58361.801067]  do_vfs_ioctl+0x42e/0x6b0
[58361.801069]  ? handle_mm_fault+0xc4/0x1f0
[58361.801070]  ksys_ioctl+0x5e/0x90
[58361.801071]  __x64_sys_ioctl+0x16/0x20
[58361.801073]  do_syscall_64+0x5b/0x1a0
[58361.801074]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[58361.801075] RIP: 0033:0x7f2fc418a25b
[58361.801077] Code: 0f 1e fa 48 8b 05 25 9c 0c 00 64 c7 00 26 00 00 00 48 c7 c0 ff ff ff ff c3 66 0f 1f 44 00 00 f3 0f 1e fa b8 10 00 00 00 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d f5 9b 0c 00 f7 d8 64 89 01 48
```

Reply:
@gamanakis According to this error @behlendorf is simply wrong and it isn't a small alloc:
`Large kmem_alloc(131272, 0x1000), please file an issue at:`

Reply:
That is bigger than I expected from initially looking at the structure.  I see that I overlooked the `->l2ad_log_blk` member which is not individually allocated like `->l2ad_dev_hdr`.  This is fine.

There are some small benefits to independently allocating a 128k `l2arc_log_blk_phys_t  l2ad_log_blk` but none are crucial.  1) vmem_alloc() uses virtual address space which is relatively limited on 32-bit systems,  2) the structure is long lived which pins that address space indefinitely, 3) the `l2arc_log_blk_phys_t` would then page aligned in memory, 4) for portability reasons there's interest in minimizing the use of `vmem_alloc()`.

Reply:
I am not sure allocating independently a `l2arc_log_blk_phys_t l2ad_log_blk` is going to help, since this would have to be done with `vmem_zalloc()`, otherwise the same error occurs (`Large kmem_alloc(131272, 0x1000)`).
I will leave this as is for now.